olm/catalogd/v1alpha1/clustercatalog_types.go

```diff
@@ -0,0 +1,223 @@
+// +enum
+type SourceType string
+
+const (
+	SourceTypeImage SourceType = "Image"
+
+	TypeProgressing = "Progressing"
+	TypeServing     = "Serving"
+
+	// Serving reasons
+	ReasonAvailable   = "Available"
+	ReasonUnavailable = "Unavailable"
+	ReasonDisabled    = "Disabled"
+
+	// Progressing reasons
+	ReasonSucceeded = "Succeeded"
+	ReasonRetrying  = "Retrying"
+	ReasonBlocked   = "Blocked"
+
+	MetadataNameLabel = "olm.operatorframework.io/metadata.name"
+
+	AvailabilityEnabled  = "Enabled"
+	AvailabilityDisabled = "Disabled"
+)
+
+//+kubebuilder:object:root=true
+//+kubebuilder:resource:scope=Cluster
+//+kubebuilder:subresource:status
+//+kubebuilder:printcolumn:name=LastUnpacked,type=date,JSONPath=`.status.lastUnpacked`
+//+kubebuilder:printcolumn:name="Serving",type=string,JSONPath=`.status.conditions[?(@.type=="Serving")].status`
+//+kubebuilder:printcolumn:name=Age,type=date,JSONPath=`.metadata.creationTimestamp`
+
+// ClusterCatalog enables users to make File-Based Catalog (FBC) catalog data available to the cluster.
+// For more information on FBC, see https://olm.operatorframework.io/docs/reference/file-based-catalogs/#docs
+type ClusterCatalog struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	Spec   ClusterCatalogSpec   `json:"spec"`
+	Status ClusterCatalogStatus `json:"status,omitempty"`
+}
+
+//+kubebuilder:object:root=true
+
+// ClusterCatalogList contains a list of ClusterCatalog
+type ClusterCatalogList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []ClusterCatalog `json:"items"`
+}
+
+// ClusterCatalogSpec defines the desired state of ClusterCatalog
+// +kubebuilder:validation:XValidation:rule="!has(self.source.image.pollInterval) || (self.source.image.ref.find('@sha256:') == \"\")",message="cannot specify PollInterval while using digest-based image"
+type ClusterCatalogSpec struct {
+	// source is a required field that allows the user to define the source of a Catalog that contains catalog metadata in the File-Based Catalog (FBC) format.
+	//
+	// Below is a minimal example of a ClusterCatalogSpec that sources a catalog from an image:
+	//
+	//  source:
+	//    type: Image
+	//    image:
+	//      ref: quay.io/operatorhubio/catalog:latest
+	//
+	// For more information on FBC, see https://olm.operatorframework.io/docs/reference/file-based-catalogs/#docs
+	Source CatalogSource `json:"source"`
+
+	// priority is an optional field that allows the user to define a priority for a ClusterCatalog.
+	// A ClusterCatalog's priority is used by clients as a tie-breaker between ClusterCatalogs that meet the client's requirements.
+	// For example, in the case where multiple ClusterCatalogs provide the same bundle.
+	// A higher number means higher priority. Negative numbers are also accepted.
+	// When omitted, the default priority is 0.
+	// +kubebuilder:default:=0
+	// +optional
+	Priority int32 `json:"priority"`
```

Comment:
Minimum would be -2147483648 since it is the largest negative number that a signed int32 can have. Maximum would be 214748647.

Reply:
Do you think users need the whole range? Is there any reason to impose bounds that aren't at the top/bottom of that range?

Wondering if there's a use case for "I must be the top most priority" and having that as something round rather than 214748647 might be easier?

Reply:
See https://github.com/openshift/api/pull/2067/files#r1819056720 for additional motivations for having an open range (I believe the same reasons apply re migration of the existing OLMv0 CatalogSource API)